src/govuk/components/accordion/accordion.mjs

```diff
@@ -322,14 +353,21 @@ Accordion.prototype.onShowOrHideAllToggle = function () {
  * Set section attributes when opened/closed
  *
  * @param {boolean} expanded - Section expanded
- * @param {HTMLElement} $section - Section element
+ * @param {Element} $section - Section element
  */
 Accordion.prototype.setExpanded = function (expanded, $section) {
   var $showHideIcon = $section.querySelector('.' + this.upChevronIconClass)
   var $showHideText = $section.querySelector('.' + this.sectionShowHideTextClass)
   var $button = $section.querySelector('.' + this.sectionButtonClass)
   var $content = $section.querySelector('.' + this.sectionContentClass)
 
+  if (!$showHideIcon ||
+    !$showHideText ||
+    !$button ||
+    !$content) {
+    return
+  }
```

Comment:
This is a good example of where I think we're fighting our types a little bit – we 'know' these elements should exist, because this class created them!

This is however muddled by the fact that theoretically this is a function that's callable from outside of the class with _anything_ as a `$section`.

Maybe it's necessary to get us to a place where we can refactor and try and eliminate this sort of thing? Should encourage us to write better code and design better APIs.

Thoughts, anyone?

Reply:
This bit of defensiveness is not really type related (only `$showHideText` makes TypeScript choke and will [have its check turned into an `instanceof` in the next PR in the chain](https://github.com/alphagov/govuk-frontend/pull/3123/files#diff-57a715a44d5c5b57c68fc41c488d014d8c974443b1233f7ef36517b4ab5aea65R368)). 

I guess this is stemming more from the fact that anything could be passed as a `$section` like you mention, but I'll let @colinrotherham confirm. This could also be for consistency to have checks on all 4 elements. 

This sparked two thoughts (questions):

1. I think we need to have a discussion on how we hande defensiveness across the library. What we defend against (eg. wrong type, unexpected/missing markup on the top of my head), how we inform users of their mistakes (I'm not sure running an early return is helpful to them) but without burdening the size of the library either.
2. Typing aside, that re-querying would be a good argument for that `AccordionSection` you mentioned to me yesterday, where we could store the things we create and have them already typed. I guess that would be a breaking change, though, as it'd change methods like the `setExpanded` this code is in.

Reply:
@romaricpascal @36degrees It's only `$showHideText` that the compiler is worried about

Because we re-query the DOM on every toggle the compiler knows `this.sectionShowHideTextClass` might have changed, so `$showHideText.innerText` needs confirming as an HTMLElement before it's happy to proceed

The `lint-types` branch will still need at least this though:

```js
if (!($showHideText instanceof HTMLElement)) {
  return
}
```

Want me to drop the new checks for this PR?

Reply:
I would be in favour of only having the `if` necessary to make the type linting pass, but won't die on that hill 😄

Reply:
Thanks! The alternative is we find all these elements on `init()`

Saves us having to re-query them on every toggle